mne/preprocessing/ecg.py

```diff
@@ -366,7 +366,24 @@ def create_ecg_epochs(raw, ch_name=None, event_id=999, picks=None, tmin=-0.5,
 
 @verbose
 def _make_ecg(inst, start, stop, reject_by_annotation=False, verbose=None):
-    """Create ECG signal from cross channel average."""
+    """Create ECG signal from cross channel average.
+
+    BUG: Used in `find_bads_ecg`, which says that start and stop
+         can be int (=sample indices) or float (=seconds). But
+         below, these are treated differently for Raw, Epochs, Evoked:
+
+         Raw: will raise TypeError if float is passed, ints are processed
+              correctly
+
+         Epochs: ints are treated as seconds, NOT sample indices, floats
+                 are processed correctly, except that `ecg` and `times`
+                 seem to be out of sync.
+
+         Evoked: start and stop are erroneously not processed at all.
+
+         See: https://github.com/mne-tools/mne-python/pull/9556#issuecomment-879106098
+
+    """  # noqa: E501
```

Comment:
I think this specific bug can be solved within `_make_ecg` by checking the type of start and stop:

- if int and Raw -> use start and stop of raw.get_data
- if float and Raw -> use the new tmin and tmax of raw.get_data (same for Epochs and Evoked, which will get their tmin/tmax)
- if int and Epochs -> some custom code to index into the _data array (same for Evoked)

`_make_ecg` is only used in `find_bads_ecg` (where the current bug is), and in `find_ecg_events`, which is affected a little differently: it always passes `start=0, stop=None` to `_make_ecg` (no customization by user via params possible)

- For Raw, this is fine --> corresponds to the full signal

For other (Epochs / Evoked) this would be buggy (see Details) ... BUT `find_ecg_events` only accepts Raw, so it's fine.

So yes, I think the proposed fix would solve all instances.


<details>

- For Epochs, this is buggy --> will get the **end** of the signal right, but the beginning is always at 0 SECONDS (because .crop is used)

- For Evoked, this is fine --> "lucky accident" corresponds to full signal because Evoked start/stop is erroneously not handled in `_make_ecg`

</details>

Reply:
proposed fix in 14a70b92cbd7041c81d73e1b4d9482e662113d1a

I just need the reverse function to `inst.time_as_idx` --> so something like `inst.idx_as_time` :wink:

Reply:
if there is something in the codebase to gracefully handle this already, I'd be thankful for pointers @larsoner ... it's a bit tricky otherwise, because times can be negative (baseline), so simple division/multiplication with sfreq won't do

Reply:
continued in https://github.com/mne-tools/mne-python/pull/9556#discussion_r670481055